pkg/ingester/ingester_activity.go

```diff
@@ -150,5 +151,74 @@ func (i *ActivityTrackerWrapper) ShutdownHandler(w http.ResponseWriter, r *http.
 func requestActivity(ctx context.Context, name string, req interface{}) string {
 	userID, _ := tenant.TenantID(ctx)
 	traceID, _ := tracing.ExtractSampledTraceID(ctx)
-	return fmt.Sprintf("%s: user=%q trace=%q request=%v", name, userID, traceID, req)
+
+	switch r := req.(type) {
+	case *client.QueryRequest:
+		// To minimize memory allocation, make use of an optimized stringer implementation
+		// for *client.QueryRequest type, as this request can be invoked multiple times per second.
+		return queryRequestActivity(name, userID, traceID, r)
+
+	default:
+		return fmt.Sprintf("%s: user=%q trace=%q request=%v", name, userID, traceID, req)
+	}
+}
+
+func queryRequestActivity(name, userID, traceID string, req *client.QueryRequest) string {
+	sb := bytes.NewBuffer(
+		make([]byte, 0, 8192),
+	)
+	sb.WriteString(name)
+	sb.WriteString(`: user="`)
+	sb.WriteString(userID)
```

Comment:
Previous version was quoting userID and traceID. We could do the same using `strconv.Quote(userID)`, to make sure that we get readable output in case of obscure values.

Reply:
Changed @ c29703598c3ca5ab478683eef9ed292528f0325b